File: snippets/design-patterns/creational/factory.py
'''
--- Factory Pattern ---
Use Case: When you want to delegate the responsibility of creating objects to a factory class.
Example: Creating objects with a common interface but different types.
'''

from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def draw(self):
        pass

class Circle(Shape):
    def draw(self):
        return "Drawing a Circle"

class Square(Shape):
    def draw(self):
        return "Drawing a Square"

class ShapeFactory:
    @staticmethod
    def get_shape(shape_type: str) -> Shape:
        if shape_type == "circle":
            return Circle()
        elif shape_type == "square":
            return Square()
        else:
            raise ValueError("Invalid shape type")

# Example Usage
circle = ShapeFactory.get_shape("circle")
print(circle.draw())  # Output: Drawing a Circle

File: snippets/design-patterns/creational/factory.go
// --- Factory Pattern ---
// Use Case: Create objects without exposing the creation logic.
// Example: Building structs with different configurations.
// Description: The Factory Pattern provides an interface for creating objects,
// allowing the creation logic to be abstracted and centralized.

package main

import "fmt"

type Shape interface {
	Draw() string
}

type Circle struct{}

func (c Circle) Draw() string { return "Drawing Circle" }

type Square struct{}

func (s Square) Draw() string { return "Drawing Square" }

func ShapeFactory(shapeType string) Shape {
	switch shapeType {
	case "circle":
		return Circle{}
	case "square":
		return Square{}
	default:
		return nil
	}
}

func main() {
	shape := ShapeFactory("circle")
	fmt.Println(shape.Draw()) // Output: Drawing Circle
}
